### DISCUSSION TITLE
Shipping firmware outside of LVFS

### DISCUSSION THREAD

**Discussion by DylanVanAssche** [fwupd/fwupd#4209] [Q&A] [answered]
**Describe the question**

I have been digging into the LVFS documentation on how to run a separate instance or a separate repository.
An instance is overkill so a simple repository would be more than enough.
However, I haven't found how to actually setup such a repository, I found a script to generate metadata: https://lvfs.readthedocs.io/en/latest/offline.html#building-a-custom-remote

So I was wondering how to do this actually if you start from some firmware files?
I already have this working:

1. Zip all firmware files (fastboot plugin expects a ZIP file)
2. Create the XML metadata for the CAB file
3. Combine everything as a CAB file

However, it seems that the CAB file needs to be signed somehow and then published in some way to make it working.
How can I proceed here?

- Signing
  - Key: I think following the steps here: https://github.com/hughsie/libjcat#public-key-signing ?
  - Signing itself? From the docs: `jcat-tool firmware.xml.gz.jcat firmware.xml.gz CERTIFICATE PRIVATE_KEY` but where do these files come from: `firmware.xml.gz.jcat firmware.xml.gz` ?
- Repository
  - Metadata: I think fwupd expects metadata to know what's in the repo. That's achieved with: https://gitlab.com/fwupd/lvfs-website/-/blob/master/local.py ?
  - Adding & trusting remote: Adding seems to be just some conf file in `/etc/fwupd/remotes.d`, but doesn't it need to be trusted in some way? I needed to set `OnlyTrusted=false` in the fwupd config with the CAB file I had above to force fwupd to flash it.

**fwupd version information**

1.7.4

**Comment by hughsie** [comment]
> the CAB file needs to be signed somehow

The can file doesn't have to be signed, but if it's not signed you need to do the `OnlyTrusted=false` thing and accept the PolicyKit auth for all files (if you're compiling with polkit that is).

> I think following the steps here:

Yup, use https://github.com/hughsie/libjcat#public-key-signing and generate a certificate and privkey with the correct data. You need to sign the things *in* the .can file, namely the firmware and the metainfo.xml file, then include the .jcat file in the archive too.

> That's achieved with: https://gitlab.com/fwupd/lvfs-website/-/blob/master/local.py ?

Yup.

> but doesn't it need to be trusted in some way

Yes, you need to install the certificate you signed the files into `/etc/pki/fwupd/` and the cert you signed the metadata into `/etc/pki/fwupd-metadata/` -- and they can be the same cert too. That's the trust side of the equation.

But, _back a bit_. If you're doing this as a contractor to a OEM, could you just use the LVFS to sign the firmware files? As I understand you're building firmware *for* the OEM as an alternate to the vendor provided firmware. We support that just fine on the LVFS now, and your user account would be restricted to only that branch for just that hardware.

**Reply by DylanVanAssche** [reply]
Thanks for explaining this!

> But, back a bit. If you're doing this as a contractor to a OEM, could you just use the LVFS to sign the firmware files? As I understand you're building firmware for the OEM as an alternate to the vendor provided firmware. We support that just fine on the LVFS now, and your user account would be restricted to only that branch for just that hardware.

Well we want to ship independent firmware which is not officially verified by the OEM or hardware vendor (ODM).
The OEM does not provide firmware even on LVFS, so I'm not sure how that would work and we (as a distro) don't want to step on the toes of the hardware vendor and the OEM by making it available on LVFS.

So the idea was to supply an additional repo with that unofficial firmware for our distro users to fix a known root exploit (which is actively being abused). The unofficial firmware has a separate branch though.

**Comment by hughsie** [comment]
> we (as a distro) don't want to step on the toes of the hardware vendor and the OEM by making it available on LVFS

Hmm, I can understand this. It would also be tricky to get the required redistribution permission, as I'm assuming you're not building the modem firmware from pristine open source code with an open toolchain.

> to fix a known root exploit

Can you share more details? This is the kind of shitposting I love to do on twitter to try to force vendors onto the LVFS.

**Reply by DylanVanAssche** [reply]
> Hmm, I can understand this. It would also be tricky to get the required redistribution permission, as I'm assuming you're not building the modem firmware from pristine open source code with an open toolchain.

Well all parts we want to ship use open stuff. Builds with Yocto. 
The only closed source part is the baseband modem firmware which we just omit from the upgrade package to make it FOSS.
The repo is here: https://github.com/Biktorgj/pinephone_modem_sdk

So the legal issues are 'worked around' by not shipping the proprietary parts.
Moreover, modifying the baseband firmware would be serious violation of laws in many countries so that's why (and the lack of source) we will never touch that part.

> Can you share more details? This is the kind of shitposting I love to do on twitter to try to force vendors onto the LVFS.

It is known, the vendor even paid a bounty: https://nns.ee/blog/2021/04/03/modem-rce.html

I wrote about everything already on my blog to make this problem go away: https://dylanvanassche.be/blog/2022/pinephone-modem-upgrade/

The OEM has an account but the uploads only on request to LVFS by the ODM. I can imagine that this a paid service, but I have no clue.
I already brought you, OEM and ODM in contact with each other before Christmas over e-mail (I sent an e-mail, never heard back, asked again, no reply).

Even though you convince to get the OEM on LVFS for 'free', this won't solve anything. The 'FOSS' variant is more stable than the original firmware.

**Reply by hughsie** [reply]
I've been talking to Lenovo about this too; I'm going to build a document making all this a lot more formal, and also making a few more exciting things possible.

**Reply by DylanVanAssche** [reply]
Oh great news! If you need anything to discuss in real-time, I'm available on Matrix as @dylanvanassche@dylanvanassche.be and can join any Matrix/IRC room if needed.

**Reply by hughsie** [reply]
I've tried to codify this in some new documentation: https://gitlab.com/fwupd/lvfs-website/-/merge_requests/1207 -- feedback welcome.

**Reply by DylanVanAssche** [reply]
Love it! Looks great!

**Comment by hughsie** [comment, marked as answer]
> The 'FOSS' variant is more stable than the original firmware

So back a bit; got a link to the FOSS firmware source code?

**Reply by DylanVanAssche** [reply]
Linked above: https://github.com/Biktorgj/pinephone_modem_sdk

**Reply by hughsie** [reply]
So, is there a way to only ship the pinephone_modem_sdk in the firmware update, or do you need to include other Quectel nonfree blobs at the same time? If the latter, is it possible to unglue them?

**Reply by DylanVanAssche** [reply]
This devices uses fastboot with various partitions (like an Android phone).
If we only ship the parts from pinephone_modem_sdk we don't touch the proprietary parts on different partitions.
This makes the package fully FOSS.

The proprietary part (ADSP firmware) is independent of the `pinephone_modem_sdk` releases. It works with any ADSP firmware release from Quectel.

**Reply by hughsie** [reply]
>  This makes the package fully FOSS.

So in the circumstances, _maybe_ it's the right thing to distribute the FOSS-only firmware for that specific modem, but restricted with DMI checks to the pinephone platform, **via the LVFS itself**? It would be the second non-vendor-approved firmware on the LVFS (we also ship the bcm57xx OSS firmware) but in the circumstances it seems like the most secure and pragmatic thing to do.

**Reply by DylanVanAssche** [reply]
That's great to hear! LVFS would be the ultimate goal of course.
The LVFS account, who would control it then? The distro, the author of the pinephone_modem_sdk, etc. ?

I asked the author again of there are absolutely no proprietary parts, he is certain and will recheck the licenses for a final time.

> DMI checks?

I'm not familiar how that would go on an ARM based platform.

**Reply by kcxt** [reply]
You could parse devicetree to ensure it's a PinePhone and not some other device.

**Reply by hughsie** [reply]
> The LVFS account, who would control it then

For BCM57xx we have a single-person vendor; the upstream maintainer of the code. It just has to be someone comfortable signing something that states the firmware is legally redistributable.

> I'm not familiar how that would go on an ARM based platform.

What does `sudo fwupdtool hwids` say on the hardware? We've got some hardware support for SMBIOS-emulation for ARM, but it depends on what DT says.

**Reply by DylanVanAssche** [reply]
> For BCM57xx we have a single-person vendor; the upstream maintainer of the code. It just has to be someone comfortable signing something that states the firmware is legally redistributable.

Okay, guess that would the author of the pinephone_modem_sdk then, he knows which firmware version should be redistributed after careful testing. He has a whole test setup for each release. Tagging him: @biktorgj

> What does sudo fwupdtool hwids say on the hardware? We've got some hardware support for SMBIOS-emulation for ARM, but it depends on what DT says.

```
pine64-pinephone:/home/dylan# fwupdtool hwids
Computer Information
--------------------
ProductName: Pine64 PinePhone (1.2)
EnclosureKind: 8

Hardware IDs
------------
not available as 'Manufacturer' unknown
not available as 'Manufacturer' unknown
not available as 'Manufacturer' unknown
not available as 'Manufacturer' unknown
not available as 'Manufacturer' unknown
not available as 'Manufacturer' unknown
not available as 'Manufacturer' unknown
not available as 'Manufacturer' unknown
not available as 'Manufacturer' unknown
not available as 'Manufacturer' unknown
not available as 'Manufacturer' unknown
not available as 'Manufacturer' unknown
not available as 'Manufacturer' unknown
not available as 'Manufacturer' unknown
not available as 'Manufacturer' unknown
```

The Pine64 PinePhone Pro also has the same modem and the firmware also works for that one.
There are different PinePhone revisions as well: 1.0, 1.1, 1.2

**Reply by hughsie** [reply]
> not available as 'Manufacturer' unknown

What's the likelihood of getting a `vendor` property for the DT shipped with the device? Or is there anything in /sys/firmware/devicetree/base that can be taken as the vendor string? A tarball of the entire exported DT would be interesting.

**Reply by kcxt** [reply]
Is the compatible string enough? https://github.com/torvalds/linux/blob/master/arch/arm64/boot/dts/allwinner/sun50i-a64-pinephone-1.0.dts#L10

**Reply by hughsie** [reply]
> Is the compatible string enough

I don't know. Looking at the docs:

    The top-level compatible property typically defines a
    compatible string for the board, and then for the SoC.
    Values always given with the most-specific first, to
    least-specific last.

I suppose in this instance, the first section of the first string is probably good enough to use as a vendor. Can someone upload `/sys/firmware/devicetree/base/compatible` so I can see the format of what the kernel sets. Thanks.

**Reply by Biktorgj** [reply]
> > The 'FOSS' variant is more stable than the original firmware
> 
> So back a bit; got a link to the FOSS firmware source code?

Hi @hughsie, all the code from the FOSS firmware is in the following repos:
Little Kernel bootloader (MIT licensed AFAIK): https://github.com/Biktorgj/quectel_lk/
Non mainline kernel: https://github.com/Biktorgj/quectel_eg25_kernel
Meta-qcom layer with all the custom applications and scripts for this modem (GPL&MIT licensed) https://github.com/Biktorgj/meta-qcom/
Yocto 3.4: https://www.yoctoproject.org/

**Reply by Biktorgj** [reply]
> I suppose in this instance, the first section of the first string is probably good enough to use as a vendor. Can someone upload `/sys/firmware/devicetree/base/compatible` so I can see the format of what the kernel sets. Thanks.

Here's the output from my two samples, a Pinephone (board version 1.2) and a Pinephone Pro
Pinephone: `pine64,pinephone-1.2pine64,pinephoneallwinner,sun50i-a64`
Pinephone Pro: `pine64,pinephone-prorockchip,rk3399`

**Reply by DylanVanAssche** [reply]
Each string is separated by `\0` FYI

**Reply by hughsie** [reply]
Can you try with https://github.com/fwupd/fwupd/pull/4212 please?

**Reply by DylanVanAssche** [reply]
I tried to build your branch in #4212 but it fails to find some symbols:

```
pine64-pinephone:~$ fwupdmgr get-devices
Error relocating /usr/lib/libfwupdplugin.so.5: xb_query_context_set_flags: symbol not found
Error relocating /usr/lib/libfwupdplugin.so.5: xb_query_context_clear: symbol not found
Error relocating /usr/lib/libfwupdplugin.so.5: xb_silo_query_first_with_context: symbol not found
Error relocating /usr/lib/libfwupdplugin.so.5: xb_value_bindings_bind_str: symbol not found
Error relocating /usr/lib/libfwupdplugin.so.5: xb_query_context_get_bindings: symbol not found
Error relocating /usr/lib/libfwupdplugin.so.5: xb_silo_query_with_context: symbol not found
```

Is there something additional I need to (re)build?

**Reply by hughsie** [reply]
> symbol not found

I think you need to compile against the installed libxmlb versions - are you compiling the subproject by accident maybe?

**Reply by DylanVanAssche** [reply]
I just compile it like always: get the ZIP file of that commit, build it and install it.
I use the packages from the distro as it is done in an `APKBUILD` file instead of compiling it 'manually' (example I modified to use your branch: https://git.alpinelinux.org/aports/tree/community/fwupd/APKBUILD#n29).
However, it is the first time this fails like this so I was wondering if another bump of some dependency was needed first.

**Reply by DylanVanAssche** [reply]
Finally managed to get things running:

```
pine64-pinephone:/home/dylan# fwupdtool hwids
Computer Information
--------------------
Manufacturer: pine64
ProductName: Pine64 PinePhone (1.2)
EnclosureKind: 8

Hardware IDs
------------
not available as 'Family' unknown
not available as 'Family' unknown
not available as 'BiosVendor' unknown
not available as 'Family' unknown
not available as 'Family' unknown
not available as 'Family' unknown
not available as 'ProductSku' unknown
not available as 'ProductSku' unknown
not available as 'BaseboardManufacturer' unknown
{e03dc566-b4d0-5e5c-8006-4d76be5a173f}   <- Manufacturer + ProductName
not available as 'Family' unknown
not available as 'Family' unknown
{d868fc81-f35c-501b-95cc-0143f94baa7b}   <- Manufacturer + EnclosureKind
not available as 'BaseboardManufacturer' unknown
{adad873f-33d3-5477-bd5e-0870eca37444}   <- Manufacturer
```

Note: you need to have the main branch of `libxmlb`. There seems to be some changes regarding symbols in there causing issues.
So a release of `libxmlb` together with this would be highly appreciated ;)


If another revision of the PinePhone is used, the `1.2` part will be either `1.0` or `1.1`

**Reply by hughsie** [reply]
_Woot_, that looks good to me. Can you add a CHID requirement to your firmware of `adad873f-33d3-5477-bd5e-0870eca37444` and then I can set up the LVFS account.

**Reply by DylanVanAssche** [reply]
Adding that shouldn't be a problem, it is just some metadata here: https://lvfs.readthedocs.io/en/latest/metainfo.html#using-chid
We can test everything with CHID as well on LVFS testing if that's okay for you?
Would need fwupd 1.7.5 or higher as well.

**Reply by hughsie** [reply]
> if that's okay for you

Certainly! Is there a link for the firmware on the LVFS now/ I can't remember if you have an account there :)

>  Would need fwupd 1.7.5 or higher as well.

Ya, I'll do a new release next week too.

**Reply by DylanVanAssche** [reply]
Firmware itself is at https://github.com/Biktorgj/pinephone_modem_sdk
The repo contains already some template metadata: https://github.com/Biktorgj/pinephone_modem_sdk/blob/honister/tools/fwupd/prototype.xml
I will add the CHID there (copy it from docs, change the GUID)

> Ya, I'll do a new release next week too.

Doesn't need to be super fast ;) Having a week or 2 would be great to make sure everything is set up properly and tested on various PinePhones (also the PinePhone Pro would need this)

**Reply by DylanVanAssche** [reply]
I don't have an account on LVFS.

**Reply by DylanVanAssche** [reply]
Oh I almost forgot: the modem firmware would be distributed by @Biktorgj 
So Biktorgj would need an LVFS account, not me ;)

However, we have plans at postmarketOS to do the same for platform firmware like U-Boot, Tow-Boot, etc. so people can easily and safely upgrade their bootloader on their phone. We're working with other distros to get this thing going see https://gitlab.com/pine64-org/platform-firmware

**Reply by hughsie** [reply]
So, setting up an account for PinePhone is pretty easy; but what's your legal relationship to them? e.g. are you a paid contractor to them, and in which case what's your company name please? Happy to discuss on email if you'd prefer.

**Reply by hughsie** [reply]
Okay, answers for Biktorgj then please :)

**Reply by DylanVanAssche** [reply]
Well none of us are a company :D we're just people who like to hack and liberate hardware for a hobby :joy:

**Reply by hughsie** [reply]
So maybe a group name of `pinephone_community`? Got a better name?

**Reply by Biktorgj** [reply]
The work I've done with the modem is not really endorsed by or related to my work in any way, just acting in my own name here really.
I don't have a legal relationship with either Pine64 or Quectel, just working for free to make things work :)

I don't really know too well how LVFS handles users and groups, as you suggested, something like "pine64_community" where different users could be linked to that group would be nice. That way it could account for other devices from Pine64 too, since they only do the hardware and leave all the software side of things to the community, and they have quite a lot of devices already

**Reply by hughsie** [reply]
@Biktorgj pine64_community works for me. Some questions:

* What would be the alternate branch name be? Something like `community` perhaps?
* Would the namespace be something like `org.pine64` I assume?
* What's the vendor ID for the modem device? (it's in the `fwupdmgr get-devices` output)
* What's your email address please?

**Reply by DylanVanAssche** [reply]
```
└─QUECTEL Mobile Broadband Module:
      Device ID:          976c4a39e87f61e6940ea6a8d39c583cfa99615f
      Summary:            Quectel EG25-G modem
      Current version:    EG25GGBR07A08M2G - 05010820
      Vendor:             QUALCOMM INCORPORATED (USB:0x2C7C)
      Release Branch:     FOSS-002
      Update Error:       Cannot install update when not on AC power
      GUIDs:              db379a33-254f-5140-b37e-d36ae7e5c039
                          c33a4560-8681-55b6-bbb6-85258f2de149 ← USB\VID_2C7C&PID_0125&REV_0318&CARRIER_ROW_GENERIC_3GPP
                          1a2996cb-f86e-5583-a464-e1b96e1c6ae9 ← USB\VID_2C7C&PID_0125&REV_0318
                          587bf468-6859-5522-93a7-6cce552a0aa3 ← USB\VID_2C7C&PID_0125
                          22ae45db-f68e-5c55-9c02-4557dca238ec ← USB\VID_2C7C
      Device Flags:       • System requires external power source
                          • Updatable

```

Not sure if @Biktorgj has the latest fwupd available, but here it is with branch support ^^^.
Currently `FOSS-002`, but can be `community`.

@hughsie So if tomorrow we have other firmware for the Pine64 devices, we can add more accounts to the group?

**Reply by hughsie** [reply]
> we can add more accounts to the group?

Yes, but you'll need to ask me to add accounts for you as there's no legal company assigned.

**Comment by Biktorgj** [comment]
@hughsie `community` works for me, namespace too :)
My contact email would be biktorgj_lvfs [at] biktorgj.com
USB Vendor ID is `0x2c7c` / Device ID is `0x0125` 
Here's the entire output of `fwupdmgr get-devices`, just in case:
```
└─QUECTEL Mobile Broadband Module:
      Device ID:          976c4a39e87f61e6940ea6a8d39c583cfa99615f
      Summary:            Quectel EG25-G modem
      Current version:    EG25GGBR07A08M2G - 05040820
      Vendor:             QUALCOMM INCORPORATED (USB:0x2C7C)
      Release Branch:     FOSS
      Update Error:       Cannot install update when not on AC power
      GUIDs:              db379a33-254f-5140-b37e-d36ae7e5c039
                          c33a4560-8681-55b6-bbb6-85258f2de149
                          1a2996cb-f86e-5583-a464-e1b96e1c6ae9
                          587bf468-6859-5522-93a7-6cce552a0aa3
                          22ae45db-f68e-5c55-9c02-4557dca238ec
      Device Flags:       • System requires external power source
                          • Updatable
    
```

**Reply by hughsie** [reply]
> community works for me, namespace too :)

Both added. I've added you an account on the LVFS now, let me know how you get on.

**Reply by Biktorgj** [reply]
> > community works for me, namespace too :)
> 
> Both added. I've added you an account on the LVFS now, let me know how you get on.

Thank you very much, just logged in!
Will tour around for a bit and reread all the documentation before uploading anything :)

**Reply by hughsie** [reply]
> reread all the documentation before uploading anything

Don't worry about getting it 100% correct -- you can edit most of the firmware metadata on the LVFS; the webservice will edit the metadata in the archive and resign it for you automatically.

**Reply by DylanVanAssche** [reply]
@hughsie Could I have also an account there? So I can help @Biktorgj out with the firmware testing and setting everything up.
I did the fwupd side of things so I'm more familiar with the metadata and other features of fwupd.

My e-mail is `account AT dylanvanassche DOT be`

Thanks!

**Reply by hughsie** [reply]
Done; you can see and modify each others firmware now too.
